internal/kafka/internal/workers/cluster_mgrs/dynamic_scale_down_mgr.go

```diff
@@ -0,0 +1,321 @@
+
+		dataplaneClusterConfig: dataplaneClusterConfig,
+		clusterProvidersConfig: clusterProvidersConfig,
+		kafkaConfig:            kafkaConfig,
+		clusterService:         clusterService,
+	}
+}
+
+func (m *DynamicScaleDownManager) Start() {
+	m.StartWorker(m)
+}
+
+func (m *DynamicScaleDownManager) Stop() {
+	m.StopWorker(m)
+}
+
+func (m *DynamicScaleDownManager) Reconcile() []error {
+	var errList fleeterrors.ErrorList
+	if !m.dataplaneClusterConfig.IsDataPlaneAutoScalingEnabled() {
+		glog.Infoln("dynamic scaling is disabled. Dynamic scale down reconcile event skipped")
+		return nil
+	}
+	glog.Infoln("running dynamic scale down reconcile event")
+	defer m.logReconcileEventEnd()
+
+	err := m.processDynamicScaleDownReconcileEvent()
+	if err != nil {
+		errList.AddErrors(err)
+	}
+
+	return errList.ToErrorSlice()
+}
+
+func (m *DynamicScaleDownManager) processDynamicScaleDownReconcileEvent() error {
+	var errList fleeterrors.ErrorList
+	kafkaStreamingUnitCountPerClusterList, err := m.clusterService.FindStreamingUnitCountByClusterAndInstanceType()
+
+	if err != nil {
+		errList.AddErrors(err)
+		return errList
+	}
+
+	type processed struct {
+		indexesOfStreamingUnitForSameClusterID []int
+		processed                              bool
+	}
+
+	var alreadyProcessedClusters map[string]processed = map[string]processed{}
+
+	for i, kafkaStreamingUnitCountPerCluster := range kafkaStreamingUnitCountPerClusterList {
+		clusterID := kafkaStreamingUnitCountPerCluster.ClusterId
+		existing, ok := alreadyProcessedClusters[clusterID]
+
+		if !ok {
+			alreadyProcessedClusters[clusterID] = processed{
+				indexesOfStreamingUnitForSameClusterID: []int{i},
+				processed:                              false,
+			}
+		} else {
+			alreadyProcessedClusters[clusterID] = processed{
+				indexesOfStreamingUnitForSameClusterID: append(existing.indexesOfStreamingUnitForSameClusterID, i),
+				processed:                              existing.processed,
+			}
+		}
+	}
+
+	for _, suCount := range kafkaStreamingUnitCountPerClusterList {
+		clusterID := suCount.ClusterId
+		existing, alreadyProcessed := alreadyProcessedClusters[clusterID]
+		if alreadyProcessed && existing.processed { // skip already processed
+			continue
+		}
+
+		alreadyProcessedClusters[clusterID] = processed{
+			indexesOfStreamingUnitForSameClusterID: existing.indexesOfStreamingUnitForSameClusterID,
+			processed:                              true,
+		}
+
+		if suCount.Status != api.ClusterReady.String() { // do not consider non ready cluster for scale down evaluation
+			continue
+		}
+
+		var regionsSupportedInstanceType config.InstanceTypeMap
+		provider, ok := m.clusterProvidersConfig.ProvidersConfig.SupportedProviders.GetByName(suCount.CloudProvider)
+		if ok {
+			region, regionFound := provider.Regions.GetByName(suCount.Region)
+			if regionFound {
+				regionsSupportedInstanceType = region.SupportedInstanceTypes
+			}
+		}
+
+		dynamicScaleDownProcessor := &standardDynamicScaleDownProcessor{
```

Comment:
Here using `dynamicScaleDownProcessor` type instead of `standardDynamicScaleDownProcessor` might be a good idea to do less changes in case dependency injection is needed in the near future (you can see scale up mgr code for an example)

Reply:
Sounds good. I'll add the type.

Reply:
done